File: internal/generator/gosdkgen/testdata/golden/fixture-nested-arrays-api/namespaces/grid.go
package namespaces

import (
	"context"
	"fmt"
	"net/url"

	"example.com/schemas/sdk/go/fixture-nested-arrays-api/runtime"
	types "example.com/schemas/types/go/fixture-nested-arrays-api"
)

// GridNamespace contains API methods for the grid namespace.
//
// Endpoints:
// - GetGrid: GetGrid endpoint.
// - GridLabels: One grid's labels as a bare list of lists, at most `limit` rows.
// - ReplaceLabels: Replace a grid's labels; the body argument is a list of lists.
// - SaveGrid: Store a grid from a request body.
type GridNamespace struct {
	client jsonClient
}

// NewGridNamespace creates a GridNamespace.
func NewGridNamespace(
	client jsonClient,
) *GridNamespace {
	return &GridNamespace{
		client: client,
	}
}

// GetGrid calls the GET /api/grids/%v endpoint.
func (n *GridNamespace) GetGrid(
	ctx context.Context,
	Id string,
) (types.GridView, error) {
	path := fmt.Sprintf("/api/grids/%v", Id)

	var requestBody any
	var out types.GridView
	if err := n.client.DoJSON(
		ctx,
		"GET",
		path,
		nil,
		requestBody,
		&out,
	); err != nil {
		var zero types.GridView
		return zero, err
	}
	return out, nil
}

type GridGridLabelsQueryParams struct {
	Limit *float64 `json:"limit,omitempty"`
}

// GridLabels One grid's labels as a bare list of lists, at most `limit` rows.
func (n *GridNamespace) GridLabels(
	ctx context.Context,
	Id string,
	query *GridGridLabelsQueryParams,
) ([][]string, error) {
	path := fmt.Sprintf("/api/grids/%v/labels", Id)
	var queryValues url.Values
	if query != nil {
		queryValues = url.Values{}
		if query.Limit != nil {
			runtime.AddQueryParam(queryValues, "limit", *query.Limit)
		}
	}

	var requestBody any
	var out [][]string
	if err := n.client.DoJSON(
		ctx,
		"GET",
		path,
		queryValues,
		requestBody,
		&out,
	); err != nil {
		var zero [][]string
		return zero, err
	}
	return out, nil
}

type GridReplaceLabelsInput struct {
	Labels [][]string `json:"labels"`
}

// ReplaceLabels Replace a grid's labels; the body argument is a list of lists.
func (n *GridNamespace) ReplaceLabels(
	ctx context.Context,
	Id string,
	input GridReplaceLabelsInput,
) (types.GridView, error) {
	path := fmt.Sprintf("/api/grids/%v/labels", Id)
	validationErrors := types.NewValidationErrors()
	// labels is an array of arrays: an inner list is never nil, reported
	// at labels[i]; an empty one is valid.
	if input.Labels == nil {
		validationErrors.AddFieldError("labels", "required", "required field")
	}
	for i, row := range input.Labels {
		if row == nil {
			validationErrors.AddFieldError(fmt.Sprintf("labels[%d]", i), "required", "required field")
		}
	}
	if validationErrors.HasErrors() {
		var zero types.GridView
		return zero, NewValidationError(validationErrors)
	}

	var requestBody any
	requestBody = input
	var out types.GridView
	if err := n.client.DoJSON(
		ctx,
		"PUT",
		path,
		nil,
		requestBody,
		&out,
	); err != nil {
		var zero types.GridView
		return zero, err
	}
	return out, nil
}

// SaveGrid Store a grid from a request body.
func (n *GridNamespace) SaveGrid(
	ctx context.Context,
	input types.SaveGridInput,
) (types.GridView, error) {
	path := "/api/grids"
	validationErrors := types.NewValidationErrors()
	appendInputValidationErrors(validationErrors, any(&input))
	if validationErrors.HasErrors() {
		var zero types.GridView
		return zero, NewValidationError(validationErrors)
	}

	var requestBody any
	requestBody = input
	var out types.GridView
	if err := n.client.DoJSON(
		ctx,
		"POST",
		path,
		nil,
		requestBody,
		&out,
	); err != nil {
		var zero types.GridView
		return zero, err
	}
	return out, nil
}

File: internal/generator/rustsdkgen/testdata/golden/fixture-nested-arrays-api/src/namespaces/grid.rs
#[allow(unused_imports)]
use serde::Serialize;


use crate::client::HttpClient;
use crate::errors::SDKError;
use crate::runtime;
use crate::types;

#[derive(Clone)]
pub struct GridNamespace {
    client: HttpClient,
}
#[derive(Debug, Clone)]
pub struct GridLabelsQueryParams {
    pub limit: Option<f64>,
}


#[derive(Debug, Clone, Serialize)]
pub struct ReplaceLabelsInput {
    pub labels: Vec<Vec<String>>,
}



impl GridNamespace {
    pub(crate) fn new(
        client: HttpClient,
    ) -> Self {
        Self {
            client,
        }
    }
    /// get_grid endpoint.
    ///
    /// - `options.timeout_ms`: optional per-request timeout override.
    /// - `options.request_hook`: optional request builder hook.
    pub async fn get_grid(
        &self,
        id: types::IdentityUUID,
        options: Option<&runtime::RequestOptions>,
    ) -> Result<types::GridView, SDKError> {
        let path = format!("/api/grids/{}", id);
        let query_params: Vec<(String, String)> = Vec::new();
        self.client
            .request_json("GET", &path, &query_params, None, options)
            .await
    }
    /// One grid's labels as a bare list of lists, at most `limit` rows.
    ///
    /// - `query`: optional query parameters for this request.
    ///
    /// - `options.timeout_ms`: optional per-request timeout override.
    /// - `options.request_hook`: optional request builder hook.
    pub async fn grid_labels(
        &self,
        id: types::IdentityUUID,
        query: Option<&GridLabelsQueryParams>,
        options: Option<&runtime::RequestOptions>,
    ) -> Result<Vec<Vec<String>>, SDKError> {
        let path = format!("/api/grids/{}/labels", id);
        let mut query_params: Vec<(String, String)> = Vec::new();
        if let Some(query) = query {
            if let Some(query_param_value) = &query.limit {
            query_params.push(("limit".to_string(), query_param_value.to_string()));
            }
        }
        self.client
            .request_json("GET", &path, &query_params, None, options)
            .await
    }
    /// Replace a grid's labels; the body argument is a list of lists.
    ///
    /// - `options.timeout_ms`: optional per-request timeout override.
    /// - `options.request_hook`: optional request builder hook.
    pub async fn replace_labels(
        &self,
        id: types::IdentityUUID,
        input: ReplaceLabelsInput,
        options: Option<&runtime::RequestOptions>,
    ) -> Result<types::GridView, SDKError> {
        let path = format!("/api/grids/{}/labels", id);
        let query_params: Vec<(String, String)> = Vec::new();
        let body = Some(serde_json::to_value(input)?);
        self.client
            .request_json("PUT", &path, &query_params, body, options)
            .await
    }
    /// Store a grid from a request body.
    ///
    /// - `options.timeout_ms`: optional per-request timeout override.
    /// - `options.request_hook`: optional request builder hook.
    pub async fn save_grid(
        &self,
        input: types::SaveGridInput,
        options: Option<&runtime::RequestOptions>,
    ) -> Result<types::GridView, SDKError> {
        let path = "/api/grids";
        let input_value = serde_json::to_value(&input)?;
        runtime::validate_input("SaveGridInput", &input_value)?;
        let query_params: Vec<(String, String)> = Vec::new();
        let body = Some(input_value);
        self.client
            .request_json("POST", &path, &query_params, body, options)
            .await
    }
}
